### ISSUE TITLE
[INTEGRATION][DAGSTER] Add capability to include input and output datasets with the run event

### ISSUE THREAD

**Issue by dominiquetipton** [OpenLineage/OpenLineage#583]
With the initial work that has been done to integrate OpenLineage with Dagster (thanks @dalinkim!), it seems like the next step should be to include a way to emit datasets along with the run events.

## Proposal:

Introduce a way to transform Dagster assets or Outputs into InputDatasets and/or OutputDatasets to improve metadata collection.

## Proposed Implementation:

I am no Dagster expert, but after working with the assets for a bit, there unfortunately doesn't seem to be a good way to automatically transform a generic [AssetMaterialization](https://docs.dagster.io/_apidocs/ops#dagster.AssetMaterialization) or [Output](https://docs.dagster.io/_apidocs/ops#dagster.Output) into an OpenLineage Dataset. However, specific OpenLineage assets can be manually created and yielded in a Dagster job/pipeline. Through the use of a wrapper/asset-generating function, a sufficiently unique AssetMaterialization with a defined structure can be yielded and later parsed into a dataset and added to the run event.

<br>

>  Through the use of a wrapper/asset-generating function, a sufficiently unique AssetMaterialization with a defined structure can be created/yielded

This function would take in an input/output dataset and create an AssetMaterialization—with an [AssetKey](https://docs.dagster.io/_apidocs/ops#dagster.AssetKey) or unique structure that can be used to identify the asset as an OpenLineage dataset. That asset could then be yielded in a Dagster job/pipeline.

> and later parsed into a dataset

An extra `elif` check can be added to the [OpenLineage-Dagster sensor](https://github.com/OpenLineage/OpenLineage/blob/8c15b9fe9dc88799d44fc3999f18a47909acb844/integration/dagster/openlineage/dagster/sensor.py#L145-L157) to detect [when assets are yielded](https://docs.dagster.io/_apidocs/execution#dagster.DagsterEventType.ASSET_MATERIALIZATION). The asset can then be determined if it is a dataset by its AssetKey/structure and if it is, it can be parsed back into an input/output dataset by reversing the logic used in the above function.

> and added to the run event

After the datasets have been parsed and are available in the sensor, they can be passed along to the emit methods so they can be added to the run event.

<br>

**Example:**

To emit a dataset, something similar to this would be added to a job/pipeline:

```python
@op
def an_op():
    output_dataset = OutputDataset(
        namespace="example_namespace",
        name="example_name"
    )

    asset = asset_generating_function(output_dataset)
    yield asset

    # or yield it directly in the function
    # asset_yielding_function(output_dataset)
```

<br>

**Limitations of this approach:**
- In order to fit with other assets yielded by Dagster, the AssetKey should be 100% human readable (i.e. no UUIDs and have a hierarchy that makes sense). 
- The created AssetKey should be idempotent. A dataset with the same namespace/name should produce the same key so dagster can relate the assets to one another. A dataset is identified by its namespace and name and the asset should be too.
- The AssetMaterialization (most likely the AssetKey) should contain a sufficiently unique identifier so other assets are not mistakenly identified as an OpenLineage dataset.
  - A good option for the AssetKey limitations could be ["OpenLineage", "InputDataset _OR_ OutputDataset _OR_ Dataset", "<dataset.namespace>", "<dataset.name>"]  
    This approach is human readable, idempotent, and is a sufficiently unique identifier (assuming other assets beginning with OpenLineage > Dataset are not being created elsewhere).

**Comment by mobuchowski** [comment]
@dominiquetipton sorry for not commenting on this earlier. 

>I am no Dagster expert, but after working with the assets for a bit, there unfortunately doesn't seem to be a good way to automatically transform a generic [AssetMaterialization](https://docs.dagster.io/_apidocs/ops#dagster.AssetMaterialization) or [Output](https://docs.dagster.io/_apidocs/ops#dagster.Output) into an OpenLineage Dataset.

I'm also not a Dagster expert, so just want to say that from OpenLineage spec perspective the most important thing is that dataset naming is matching the naming scheme: https://github.com/OpenLineage/OpenLineage/blob/main/spec/Naming.md

This allows us to track dataset lineage between different systems, not only Dagster. Perhaps there is something that could "guide" those manually created assets to a proper structure?

**Comment by dominiquetipton** [comment]
No worries, thanks for getting around to this.

I absolutely agree. With the idea I had for this, the responsibility of following the naming scheme would fall on the user since they will be the ones creating the datasets. The only thing this proposal would do with names is prepend a categorizing/identifying prefix to the created asset's key.

With this approach the OpenLineage assets would be kept separate from the other assets. For example in the Dagster UI, the OL assets would be under a 'OpenLineage > Datasets > {dataset.namespace} > {dataset.name}' drill down menu. Also, the sensor would only use the prefix for assisting with identifying the appropriate assets. It would send the dataset as it was originally created by the user.

**Comment by mobuchowski** [comment]
>With the idea I had for this, the responsibility of following the naming scheme would fall on the user since they will be the ones creating the datasets.

That makes sense - I just had an idea to somehow guide people to choosing good name. For example, provide base class `OutputDataset` - that has several subclasses like `BigQueryOutputDataset` that in constructor accepts GCP project, schema name and table name. This would convert those to proper dataset naming.

**Comment by dominiquetipton** [comment]
Ah, I see. While I do think that's a great idea (there have been a lot of recent discussions going on in my workplace about standardizing dataset naming conventions), I feel that may fit better with a contribution to the [client](https://github.com/OpenLineage/OpenLineage/blob/main/client/python/openlineage/client/run.py) (or a client-extention type of package) instead of in this Dagster specific issue.

**Comment by mobuchowski** [comment]
@dominiquetipton want to revisit this?

**Comment by dominiquetipton** [comment]
@mobuchowski no, unfortunately my team has shifted its focus to different areas for the time being.

**Comment by kacpermuda** [comment]
Since dagster integration has been removed from the repository, will close this issue.
